### PR TITLE
client: add CreateWithOptions function to client.KeysAPI

### PR DESCRIPTION
**Pull request by chirino**
Allows you to optionally Create nodes with:
* A TTL
* In order generated keys
* Ignore existing node values

### CODE REVIEW

**Comment by chirino** [comment]
@bcwaldon Ok this new version of the patch now uses a new func.

**Comment by chirino** [comment]
@bcwaldon ping.

File: client/keys.go [hunk 1]
```diff
@@ -103,9 +103,12 @@ type KeysAPI interface {
 	// conditions in an DeleteOptions object.
 	Delete(ctx context.Context, key string, opts *DeleteOptions) (*Response, error)
 
-	// Create is an alias for Set w/ PrevExist=false
+	// Used to Create a Node.  Same as CreateWithOptions(ctx, key, value, nil)
```
Comment (bcwaldon) [inline client/keys.go]: From https://golang.org/doc/effective_go.html: 

```
Doc comments work best as complete sentences, which allow a wide variety of automated presentations. The first sentence should be a one-sentence summary that starts with the name being declared.
```

**Comment by bcwaldon** [comment]
@chirino I talked to the etcd devs a bit offline about this and we came to the following conclusion:

The Create operation is effectively a Set w/ a requirement that the key not already exist (i.e. prevExist=false). A Create-In-Order operation does not behave like a Set - it offloads the responsibility of picking the key to the server. Trying to share an interface between these two methods is only going to confuse users of this library and make the testing matrix more complex for the existing code. Given that Create == Set but Create-In-Order != Set, the functionality you are introducing needs to be implemented independently of the Set codepath. This means a new `CreateInOrder` (actually, `CreateSeq` is preferred) method on the KeysAPI and a new method in the HTTP implementation. This new `CreateSeq` function can have a `CreateSeqOptions` struct if necessary, but it clearly should not expose any functionality that is intended for a normal Create/Set.

**Comment by chirino** [comment]
What interface is it that we are sharing?  Are you saying the CreateWithOptions is exposing functionality that is not valid for a Create?  If so what is it?

**Comment by bcwaldon** [comment]
@chirino The shared interface I'm referring to is the concept of "Create". A Create and a Create-In-Order are conceptually different operations. `CreateWithOptions(key, value, CreateOptions{InOrder: true})` is totally different than `CreateWithOptions(key, value, CreateOptions{InOrder: false})`, so I'd like to see the distinction made in the method name, not a boolean.

**Comment by chirino** [comment]
I don't agree they are different operations.  They are both create operations. 

How would you create a non-ordered node with a TTL?  Right with my change it would be something like: 
`CreateWithOptions(key, value, CreateOptions{TTL: 3000})`?

Are you saying keep the CreateWithOptions func and also add a CreateInOrder function?  

What if you want to set the TTL on for an in-order node?  Should we also have a CreateInOrderWithOptions func?

**Comment by bcwaldon** [comment]
Before we can talk about code, we need to agree that these are fundamentally different operations. We cannot agree on an API until we understand what functionality we're trying to expose. So here's how I see it:

### Create

- node will not be created unless another node does not exist at provided key (Compare-and-Swap based on existence)
- user decides full key at-which value will be written
- user provides value
- user may provide TTL

### Create-In-Order

- there is no comparison operation, like there is in a Create (CAS based on existence)
- user does not decide full key at-which value will be written
- user provides value
- user may provide TTL

As you can see, there are two incredibly important differences: 
- Create is a CAS operation, while Create-In-Order is not
- Create requires the user to provide a full key, while Create-In-Order does not

Given how important those differences are, I am not willing to accept an API that distinguishes between them using a boolean. That approach hides a *very* important decision behind a seemingly innocuous flag.

To your question about how you would "create a non-ordered node with a TTL":

```
KeysAPI.Set("/foo", "val", SetOptions{TTL: time.Second, PrevExist: true})
```

So, I do not see *any* changes necessary to the existing code, only additions. I don't know if this actually works, but this is effectively what I want to see:

```
commit a4ab23891aef11cded81210b10bfb0e745ffaa45
Author: Brian Waldon <brian@waldon.cc>
Date:   Fri Mar 20 09:07:55 2015 -0700

    client: rough CreateInOrder operation

diff --git a/client/keys.go b/client/keys.go
index 71852a6..0a88bfd 100644
--- a/client/keys.go
+++ b/client/keys.go
@@ -109,6 +109,11 @@ type KeysAPI interface {
 	// Update is an alias for Set w/ PrevExist=true
 	Update(ctx context.Context, key, value string) (*Response, error)
 
+	// CreateInOrder creates a new node in the given directory. The server
+	// chooses a unique key in the provided directory, returning the created
+	// Node in the Response.
+	CreateInOrder(ctx context.Context, dir, value string, opts *CreateInOrderOptions) (*Response, error)
+
 	// Watcher builds a new Watcher targeted at a specific Node identified
 	// by the given key. The Watcher may be configured at creation time
 	// through a WatcherOptions object. The returned Watcher is designed
@@ -161,6 +166,14 @@ type SetOptions struct {
 	TTL time.Duration
 }
 
+type CreateInOrderOptions struct {
+	// TTL defines a period of time after-which the Node should
+	// expire and no longer exist. Values <= 0 are ignored. Given
+	// that the zero-value is ignored, TTL cannot be used to set
+	// a TTL of 0.
+	TTL time.Duration
+}
+
 type GetOptions struct {
 	// Recursive defines whether or not all children of the Node
 	// should be returned.
@@ -289,6 +302,25 @@ func (k *httpKeysAPI) Update(ctx context.Context, key, val string) (*Response, e
 	return k.Set(ctx, key, val, &SetOptions{PrevExist: PrevExist})
 }
 
+func (k *httpKeysAPI) CreateInOrder(ctx context.Context, dir, val string, opts *CreateInOrderOptions) (*Response, error) {
+	act := &createInOrderAction{
+		Prefix: k.prefix,
+		Dir:    dir,
+		Value:  val,
+	}
+
+	if opts != nil {
+		act.TTL = opts.TTL
+	}
+
+	resp, body, err := k.client.Do(ctx, act)
+	if err != nil {
+		return nil, err
+	}
+
+	return unmarshalHTTPResponse(resp.StatusCode, resp.Header, body)
+}
+
 func (k *httpKeysAPI) Delete(ctx context.Context, key string, opts *DeleteOptions) (*Response, error) {
 	act := &deleteAction{
 		Prefix: k.prefix,
@@ -454,6 +486,28 @@ func (a *setAction) HTTPRequest(ep url.URL) *http.Request {
 	return req
 }
 
+type createInOrderAction struct {
+	Prefix string
+	Dir    string
+	Value  string
+	TTL    time.Duration
+}
+
+func (a *createInOrderAction) HTTPRequest(ep url.URL) *http.Request {
+	u := v2KeysURL(ep, a.Prefix, a.Dir)
+	form := url.Values{}
+	form.Add("value", a.Value)
+	if a.TTL > 0 {
+		form.Add("ttl", strconv.FormatUint(uint64(a.TTL.Seconds()), 10))
+	}
+	body := strings.NewReader(form.Encode())
+
+	req, _ := http.NewRequest("POST", u.String(), body)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	return req
+}
+
 type deleteAction struct {
 	Prefix    string
 	Key       string
```

**Comment by ecnahc515** [comment]
I would really like to see this move forward. @chirino are you willing to make the changes required, or do you want someone else to pick this up?

**Comment by ecnahc515** [comment]
Alternatively I've gone ahead and made those changes in my own fork, since I'm implementing something using the new etcd api which uses this functionality:
https://github.com/coreos/etcd/compare/master...ecnahc515:create_options_inorder_api

/cc @bcwaldon

**Comment by yichengq** [comment]
@ecnahc515 has added `CreateInOrder` to handle `In order generated keys`. The `A TTL` or `Ignore existing node values` should use `Set` to make it.
Thanks for the contribution!

**Pull request closed without merging**